### ISSUE TITLE
Decoder[A].validate is subject to infinite recursion

### ISSUE THREAD

**Issue by sirocchj** [travisbrown/circe#396]
```
import io.circe._
implicit val intDecoder = Decoder[Int].validate(_ => true, "")
Json.fromInt(42).as[Int]
```

Issue is easy-peasy to fix, apply() is recursively calling itself instead of self type'd apply, submitting PR

**Comment by crakjie** [comment]
Tanks to open a pr 👍

**Comment by travisbrown** [comment]
Resolved in #396.
